File: src/geometry/dcel.ts
/**
 * DCEL - Doubly Connected Edge List
 * 
 * A data structure for representing planar subdivisions.
 * Enables efficient traversal of faces (enclosed regions).
 * 
 * Designed for Rust/WASM portability - no DOM dependencies.
 */

import type { Point } from './types';

/** Unique identifier for DCEL elements */
export type VertexId = number;
export type HalfEdgeId = number;
export type FaceId = number;

/** A vertex in the DCEL */
export interface DCELVertex {
    id: VertexId;
    x: number;
    y: number;
    /** Any half-edge originating from this vertex */
    incidentEdge: HalfEdgeId | null;
}

/** A half-edge in the DCEL */
export interface DCELHalfEdge {
    id: HalfEdgeId;
    /** Origin vertex of this half-edge */
    origin: VertexId;
    /** Twin half-edge (same edge, opposite direction) */
    twin: HalfEdgeId;
    /** Next half-edge around the face (counter-clockwise) */
    next: HalfEdgeId;
    /** Previous half-edge around the face */
    prev: HalfEdgeId;
    /** Face to the left of this half-edge */
    face: FaceId | null;
}

/** A face (enclosed region) in the DCEL */
export interface DCELFace {
    id: FaceId;
    /** One half-edge on the outer boundary */
    outerComponent: HalfEdgeId | null;
    /** Half-edges on inner boundaries (holes) */
    innerComponents: HalfEdgeId[];
    /** Whether this is the unbounded outer face */
    isOuterFace: boolean;
}

/** The complete DCEL data structure */
export class DCEL {
    vertices: Map<VertexId, DCELVertex> = new Map();
    halfEdges: Map<HalfEdgeId, DCELHalfEdge> = new Map();
    faces: Map<FaceId, DCELFace> = new Map();

    private nextVertexId = 0;
    private nextHalfEdgeId = 0;
    private nextFaceId = 0;

    /** Create a new vertex */
    createVertex(x: number, y: number): VertexId {
        const id = this.nextVertexId++;
        this.vertices.set(id, {
            id,
            x,
            y,
            incidentEdge: null,
        });
        return id;
    }

    /** Create a pair of twin half-edges between two vertices */
    createEdgePair(originA: VertexId, originB: VertexId): [HalfEdgeId, HalfEdgeId] {
        const idA = this.nextHalfEdgeId++;
        const idB = this.nextHalfEdgeId++;

        const halfEdgeA: DCELHalfEdge = {
            id: idA,
            origin: originA,
            twin: idB,
            next: idA, // Will be updated during face extraction
            prev: idA,
            face: null,
        };

        const halfEdgeB: DCELHalfEdge = {
            id: idB,
            origin: originB,
            twin: idA,
            next: idB,
            prev: idB,
            face: null,
        };

        this.halfEdges.set(idA, halfEdgeA);
        this.halfEdges.set(idB, halfEdgeB);

        // Update incident edges on vertices
        const vertexA = this.vertices.get(originA)!;
        const vertexB = this.vertices.get(originB)!;
        if (vertexA.incidentEdge === null) vertexA.incidentEdge = idA;
        if (vertexB.incidentEdge === null) vertexB.incidentEdge = idB;

        return [idA, idB];
    }

    /** Create a new face */
    createFace(outerComponent: HalfEdgeId | null, isOuterFace: boolean = false): FaceId {
        const id = this.nextFaceId++;
        this.faces.set(id, {
            id,
            outerComponent,
            innerComponents: [],
            isOuterFace,
        });
        return id;
    }

    /** Get all half-edges originating from a vertex, sorted by angle */
    getOutgoingEdges(vertexId: VertexId): HalfEdgeId[] {
        const edges: HalfEdgeId[] = [];

        for (const [id, edge] of this.halfEdges) {
            if (edge.origin === vertexId) {
                edges.push(id);
            }
        }

        // Sort by angle
        const vertex = this.vertices.get(vertexId)!;
        edges.sort((a, b) => {
            const edgeA = this.halfEdges.get(a)!;
            const edgeB = this.halfEdges.get(b)!;
            const targetA = this.vertices.get(this.halfEdges.get(edgeA.twin)!.origin)!;
            const targetB = this.vertices.get(this.halfEdges.get(edgeB.twin)!.origin)!;

            const angleA = Math.atan2(targetA.y - vertex.y, targetA.x - vertex.x);
            const angleB = Math.atan2(targetB.y - vertex.y, targetB.x - vertex.x);

            return angleA - angleB;
        });

        return edges;
    }

    /** 
     * Set up next/prev pointers based on angular ordering.
     * Call this after all edges are added.
     */
    computeNextPrevPointers(): void {
        // For each vertex, get outgoing edges sorted by angle
        for (const [vertexId] of this.vertices) {
            const outgoing = this.getOutgoingEdges(vertexId);

            if (outgoing.length === 0) continue;

            // For each outgoing edge, its twin's next is the next outgoing edge (CCW)
            for (let i = 0; i < outgoing.length; i++) {
                const currentOutgoing = outgoing[i];
                const nextOutgoing = outgoing[(i + 1) % outgoing.length];

                // The twin of currentOutgoing arrives at this vertex
                // Its next should be nextOutgoing (leaving this vertex)
                const incoming = this.halfEdges.get(currentOutgoing)!.twin;
                const incomingEdge = this.halfEdges.get(incoming)!;
                incomingEdge.next = nextOutgoing;

                // And nextOutgoing's prev is incoming
                const nextEdge = this.halfEdges.get(nextOutgoing)!;
                nextEdge.prev = incoming;
            }
        }
    }

    /**
     * Extract all faces by traversing half-edge cycles.
     * Call after computeNextPrevPointers().
     */
    extractFaces(): void {
        const visited = new Set<HalfEdgeId>();

        for (const [edgeId] of this.halfEdges) {
            if (visited.has(edgeId)) continue;

            // Traverse the face boundary
            const boundary: HalfEdgeId[] = [];
            let current = edgeId;

            do {
                if (visited.has(current)) break;
                visited.add(current);
                boundary.push(current);
                current = this.halfEdges.get(current)!.next;
            } while (current !== edgeId);

            if (boundary.length < 3) continue;

            // Create face
            const faceId = this.createFace(boundary[0], false);

            // Assign face to all half-edges in this cycle
            for (const heId of boundary) {
                this.halfEdges.get(heId)!.face = faceId;
            }
        }

        // Identify the outer (unbounded) face
        this.identifyOuterFace();
    }

    /** Identify the unbounded outer face by finding the one with largest negative area */
    private identifyOuterFace(): void {
        let outerFaceId: FaceId | null = null;
        let minArea = Infinity;

        for (const [faceId, face] of this.faces) {
            if (face.outerComponent === null) continue;

            const area = this.computeFaceArea(faceId);

            // The outer face has the most negative (clockwise) area
            if (area < minArea) {
                minArea = area;
                outerFaceId = faceId;
            }
        }

        if (outerFaceId !== null) {
            this.faces.get(outerFaceId)!.isOuterFace = true;
        }
    }

    /** Compute signed area of a face (positive = CCW, negative = CW) */
    computeFaceArea(faceId: FaceId): number {
        const face = this.faces.get(faceId);
        if (!face || face.outerComponent === null) return 0;

        let area = 0;
        let current = face.outerComponent;
        const start = current;

        do {
            const edge = this.halfEdges.get(current)!;
            const origin = this.vertices.get(edge.origin)!;
            const target = this.vertices.get(this.halfEdges.get(edge.twin)!.origin)!;

            area += origin.x * target.y - target.x * origin.y;

            current = edge.next;
        } while (current !== start);

        return area / 2;
    }

    /** Get the polygon points for a face */
    getFacePolygon(faceId: FaceId): Point[] {
        const face = this.faces.get(faceId);
        if (!face || face.outerComponent === null) return [];

        const points: Point[] = [];
        let current = face.outerComponent;
        const start = current;

        do {
            const edge = this.halfEdges.get(current)!;
            const origin = this.vertices.get(edge.origin)!;
            points.push({ x: origin.x, y: origin.y });
            current = edge.next;
        } while (current !== start);

        return points;
    }

    /** Get all bounded (non-outer) faces as polygons */
    getBoundedFaces(): Point[][] {
        const result: Point[][] = [];

        for (const [faceId, face] of this.faces) {
            if (face.isOuterFace) continue;

            const polygon = this.getFacePolygon(faceId);
            if (polygon.length >= 3) {
                result.push(polygon);
            }
        }

        return result;
    }
}

/** Find or create a vertex at the given position, with snapping tolerance */
export function findOrCreateVertex(
    dcel: DCEL,
    x: number,
    y: number,
    tolerance: number = 0.001
): VertexId {
    // Check for existing vertex within tolerance
    for (const [id, vertex] of dcel.vertices) {
        const dx = vertex.x - x;
        const dy = vertex.y - y;
        if (dx * dx + dy * dy <= tolerance * tolerance) {
            return id;
        }
    }

    // Create new vertex
    return dcel.createVertex(x, y);
}

File: wasm/src/dcel.rs
//! DCEL - Doubly Connected Edge List
//!
//! A data structure for representing planar subdivisions.
//! Enables efficient traversal of faces (enclosed regions).

use std::collections::{HashMap, HashSet};
use crate::types::Point;

/// Unique identifiers for DCEL elements
pub type VertexId = usize;
pub type HalfEdgeId = usize;
pub type FaceId = usize;

/// A vertex in the DCEL
#[derive(Debug, Clone)]
pub struct DCELVertex {
    pub id: VertexId,
    pub x: f64,
    pub y: f64,
    /// Any half-edge originating from this vertex
    pub incident_edge: Option<HalfEdgeId>,
}

/// A half-edge in the DCEL
#[derive(Debug, Clone)]
pub struct DCELHalfEdge {
    pub id: HalfEdgeId,
    /// Origin vertex of this half-edge
    pub origin: VertexId,
    /// Twin half-edge (same edge, opposite direction)
    pub twin: HalfEdgeId,
    /// Next half-edge around the face (counter-clockwise)
    pub next: HalfEdgeId,
    /// Previous half-edge around the face
    pub prev: HalfEdgeId,
    /// Face to the left of this half-edge
    pub face: Option<FaceId>,
}

/// A face (enclosed region) in the DCEL
#[derive(Debug, Clone)]
pub struct DCELFace {
    pub id: FaceId,
    /// One half-edge on the outer boundary
    pub outer_component: Option<HalfEdgeId>,
    /// Whether this is the unbounded outer face
    pub is_outer_face: bool,
}

/// The complete DCEL data structure
#[derive(Debug)]
pub struct DCEL {
    pub vertices: HashMap<VertexId, DCELVertex>,
    pub half_edges: HashMap<HalfEdgeId, DCELHalfEdge>,
    pub faces: HashMap<FaceId, DCELFace>,
    
    next_vertex_id: VertexId,
    next_half_edge_id: HalfEdgeId,
    next_face_id: FaceId,
}

impl DCEL {
    pub fn new() -> Self {
        Self {
            vertices: HashMap::new(),
            half_edges: HashMap::new(),
            faces: HashMap::new(),
            next_vertex_id: 0,
            next_half_edge_id: 0,
            next_face_id: 0,
        }
    }
    
    /// Create a new vertex
    pub fn create_vertex(&mut self, x: f64, y: f64) -> VertexId {
        let id = self.next_vertex_id;
        self.next_vertex_id += 1;
        
        self.vertices.insert(id, DCELVertex {
            id,
            x,
            y,
            incident_edge: None,
        });
        
        id
    }
    
    /// Create a pair of twin half-edges between two vertices
    pub fn create_edge_pair(&mut self, origin_a: VertexId, origin_b: VertexId) -> (HalfEdgeId, HalfEdgeId) {
        let id_a = self.next_half_edge_id;
        let id_b = self.next_half_edge_id + 1;
        self.next_half_edge_id += 2;
        
        let half_edge_a = DCELHalfEdge {
            id: id_a,
            origin: origin_a,
            twin: id_b,
            next: id_a, // Will be updated during face extraction
            prev: id_a,
            face: None,
        };
        
        let half_edge_b = DCELHalfEdge {
            id: id_b,
            origin: origin_b,
            twin: id_a,
            next: id_b,
            prev: id_b,
            face: None,
        };
        
        self.half_edges.insert(id_a, half_edge_a);
        self.half_edges.insert(id_b, half_edge_b);
        
        // Update incident edges on vertices
        if let Some(vertex_a) = self.vertices.get_mut(&origin_a) {
            if vertex_a.incident_edge.is_none() {
                vertex_a.incident_edge = Some(id_a);
            }
        }
        if let Some(vertex_b) = self.vertices.get_mut(&origin_b) {
            if vertex_b.incident_edge.is_none() {
                vertex_b.incident_edge = Some(id_b);
            }
        }
        
        (id_a, id_b)
    }
    
    /// Create a new face
    pub fn create_face(&mut self, outer_component: Option<HalfEdgeId>, is_outer_face: bool) -> FaceId {
        let id = self.next_face_id;
        self.next_face_id += 1;
        
        self.faces.insert(id, DCELFace {
            id,
            outer_component,
            is_outer_face,
        });
        
        id
    }
    
    /// Get all half-edges originating from a vertex, sorted by angle
    pub fn get_outgoing_edges(&self, vertex_id: VertexId) -> Vec<HalfEdgeId> {
        let mut edges: Vec<HalfEdgeId> = self.half_edges
            .values()
            .filter(|e| e.origin == vertex_id)
            .map(|e| e.id)
            .collect();
        
        // Sort by angle
        if let Some(vertex) = self.vertices.get(&vertex_id) {
            edges.sort_by(|&a, &b| {
                let target_a = self.get_edge_target(a);
                let target_b = self.get_edge_target(b);
                
                let angle_a = (target_a.y - vertex.y).atan2(target_a.x - vertex.x);
                let angle_b = (target_b.y - vertex.y).atan2(target_b.x - vertex.x);
                
                angle_a.partial_cmp(&angle_b).unwrap_or(std::cmp::Ordering::Equal)
            });
        }
        
        edges
    }
    
    /// Get the target vertex coordinates of a half-edge
    fn get_edge_target(&self, edge_id: HalfEdgeId) -> Point {
        let edge = self.half_edges.get(&edge_id).unwrap();
        let twin = self.half_edges.get(&edge.twin).unwrap();
        let target = self.vertices.get(&twin.origin).unwrap();
        Point::new(target.x, target.y)
    }
    
    /// Set up next/prev pointers based on angular ordering
    pub fn compute_next_prev_pointers(&mut self) {
        let vertex_ids: Vec<VertexId> = self.vertices.keys().copied().collect();
        
        for vertex_id in vertex_ids {
            let outgoing = self.get_outgoing_edges(vertex_id);
            
            if outgoing.is_empty() {
                continue;
            }
            
            // For each outgoing edge, its twin's next is the next outgoing edge (CCW)
            for i in 0..outgoing.len() {
                let current_outgoing = outgoing[i];
                let next_outgoing = outgoing[(i + 1) % outgoing.len()];
                
                // The twin of current_outgoing arrives at this vertex
                let incoming = self.half_edges.get(&current_outgoing).unwrap().twin;
                
                // Update next pointer on incoming edge
                if let Some(incoming_edge) = self.half_edges.get_mut(&incoming) {
                    incoming_edge.next = next_outgoing;
                }
                
                // Update prev pointer on next edge
                if let Some(next_edge) = self.half_edges.get_mut(&next_outgoing) {
                    next_edge.prev = incoming;
                }
            }
        }
    }
    
    /// Extract all faces by traversing half-edge cycles
    pub fn extract_faces(&mut self) {
        let mut visited: HashSet<HalfEdgeId> = HashSet::new();
        let edge_ids: Vec<HalfEdgeId> = self.half_edges.keys().copied().collect();
        
        for edge_id in edge_ids {
            if visited.contains(&edge_id) {
                continue;
            }
            
            // Traverse the face boundary
            let mut boundary = Vec::new();
            let mut current = edge_id;
            
            loop {
                if visited.contains(&current) {
                    break;
                }
                visited.insert(current);
                boundary.push(current);
                
                current = self.half_edges.get(&current).unwrap().next;
                if current == edge_id {
                    break;
                }
            }
            
            if boundary.len() < 3 {
                continue;
            }
            
            // Create face
            let face_id = self.create_face(Some(boundary[0]), false);
            
            // Assign face to all half-edges in this cycle
            for he_id in &boundary {
                if let Some(edge) = self.half_edges.get_mut(he_id) {
                    edge.face = Some(face_id);
                }
            }
        }
        
        // Identify the outer (unbounded) face
        self.identify_outer_face();
    }
    
    /// Identify the unbounded outer face by finding the one with largest negative area
    fn identify_outer_face(&mut self) {
        let mut outer_face_id: Option<FaceId> = None;
        let mut min_area = f64::INFINITY;
        
        let face_ids: Vec<FaceId> = self.faces.keys().copied().collect();
        
        for face_id in face_ids {
            let area = self.compute_face_area(face_id);
            if area < min_area {
                min_area = area;
                outer_face_id = Some(face_id);
            }
        }
        
        if let Some(id) = outer_face_id {
            if let Some(face) = self.faces.get_mut(&id) {
                face.is_outer_face = true;
            }
        }
    }
    
    /// Compute signed area of a face
    pub fn compute_face_area(&self, face_id: FaceId) -> f64 {
        let face = match self.faces.get(&face_id) {
            Some(f) => f,
            None => return 0.0,
        };
        
        let start = match face.outer_component {
            Some(s) => s,
            None => return 0.0,
        };
        
        let mut area = 0.0;
        let mut current = start;
        
        loop {
            let edge = self.half_edges.get(&current).unwrap();
            let origin = self.vertices.get(&edge.origin).unwrap();
            let twin = self.half_edges.get(&edge.twin).unwrap();
            let target = self.vertices.get(&twin.origin).unwrap();
            
            area += origin.x * target.y - target.x * origin.y;
            
            current = edge.next;
            if current == start {
                break;
            }
        }
        
        area / 2.0
    }
    
    /// Get the polygon points for a face
    pub fn get_face_polygon(&self, face_id: FaceId) -> Vec<Point> {
        let face = match self.faces.get(&face_id) {
            Some(f) => f,
            None => return Vec::new(),
        };
        
        let start = match face.outer_component {
            Some(s) => s,
            None => return Vec::new(),
        };
        
        let mut points = Vec::new();
        let mut current = start;
        
        loop {
            let edge = self.half_edges.get(&current).unwrap();
            let origin = self.vertices.get(&edge.origin).unwrap();
            points.push(Point::new(origin.x, origin.y));
            
            current = edge.next;
            if current == start {
                break;
            }
        }
        
        points
    }
}

/// Find or create a vertex at the given position, with snapping tolerance
pub fn find_or_create_vertex(dcel: &mut DCEL, x: f64, y: f64, tolerance: f64) -> VertexId {
    let tolerance_sq = tolerance * tolerance;
    let mut best_dist_sq = tolerance_sq;
    let mut best_id: Option<VertexId> = None;
    
    for (id, vertex) in &dcel.vertices {
        let dx = vertex.x - x;
        let dy = vertex.y - y;
        let dist_sq = dx * dx + dy * dy;
        
        if dist_sq <= best_dist_sq {
            best_dist_sq = dist_sq;
            best_id = Some(*id);
            
            // Optimization: if exact match, stop early
            if dist_sq < 1e-10 {
                break;
            }
        }
    }
    
    best_id.unwrap_or_else(|| dcel.create_vertex(x, y))
}

/// Check if an edge already exists between two vertices
pub fn edge_exists(dcel: &DCEL, v1: VertexId, v2: VertexId) -> bool {
    for edge in dcel.half_edges.values() {
        let twin = dcel.half_edges.get(&edge.twin).unwrap();
        if (edge.origin == v1 && twin.origin == v2) || (edge.origin == v2 && twin.origin == v1) {
            return true;
        }
    }
    false
}
